### PR TITLE
KOGITO-8120: Dashbuilder: New Map Component

### PR DESCRIPTION
**Pull request by jesuino**
Adding a new map external component. It is different from the built in one for many different reasons:

* Bubbles are now based on latitude/long - the built it component was limited to the geography;
* Users can deeply customize the map style. They can change background, fill, stroke, hover behavior and even the projection;
* Any Geo JSON file can be used and users can customize which geo key will be used for the map paint

The new component is based on [react-simple-maps](https://www.react-simple-maps.io/) and it is called "map". The basic usage is:

```
datasets:
- uuid: countries
  content: >-
            [
                  ["Brazil", "212"],
                  ["United States of America", "300"],
                  ["Canada", "38"],
                  ["Uruguay", "10"],
                  ["Russia", "120"],
                  ["China", "1200"],
                  ["India", "1000"],
                  ["Australia", "25"]
                ]
pages:
- components:
    - settings:
        component: map
        lookup:
          uuid: countries
```

It should render the following map:

image

All the supported parameters are:

### Map parameteres

* projection: The projection as described in [react-simple-maps](https://www.react-simple-maps.io/docs/composable-map/)
* background: The map background color
* geoUrl: A link to a custom geo JSON form. Default is a world map in medium resolution.

### Zoom parameters

* enableZoom: Boolean property to enable zoom. Default is false.
* zoom: Number property to define the default zoom;
* maxZoom: A number with the max allowed zoom 
* minZoom: A number with the min allowed zoom 
* zoomCenterLong: A number with the initial longitude
* zoomCenterLat: A number with the initial latitude

### Style parameters

* fill: the fill color for the geometry forms
* stroke: stroke color for the geometry forms
* strokeWidth: Number with stroke width for the geometry forms
* hover: geometry hover fill
* hoverStrokeWidth: geometry hover stroke width
* hoverStrokeFill: geometry hover stroke fill

### Bubble parameters

When using three columns with latitude, longitude the value, then bubbles are drawn. Bubbles can be customized using the following parameters:

* bubbleFill: Bubble color
* bubbleStroke: Bubble stroke color
* bubbleStrokeWidth: bubble stroke width
* bubbleOpacity: Bubble opacity
* maxBubbleSize: Bubble size when the value equals to max. The allowed value is a positive integer
* MinBubbleSize: Bubble size when the value equals to min. The allowed value is a positive integer

### Fill parameters

When using two columns with geography id and value the geography is painted according to the value. The painting can be customized using the following parameters:

* fillBegin: The color for values = min. The begin and end fill values are used in a scale from min to max. By default we use a red-ish pallete.
* fillEnd: The color for values = min.  
* geoKey: By default the map component tries to match the geography with the dataset using the geography name or the geography id. Users can customize to match with other properties using this parameter.

### Testing:

Build until dashbuilder runtime. Open dashbuilder runtime on a browser and using the web console call:

Build:
```
pnpm -F dashbuilder... build:dev
cd packages/dashbuilder/dashbuilder-runtime-parent/dashbuilder-runtime-client/target/dashbuilder-runtime-client
http-server -p 3000 --cors
```
Call to render:
```
window.setDashbuilderContent(`datasets:
- uuid: countries
  content: >-
            [
                  ["Brazil", "212"],
                  ["United States of America", "300"],
                  ["Canada", "38"],
                  ["Uruguay", "10"],
                  ["Russia", "120"],
                  ["China", "1200"],
                  ["India", "1000"],
                  ["Australia", "25"]
                ]
pages:
- components:
    - settings:
        component: map
        lookup:
          uuid: countries`)
```
Alternatively you can build until dashbuilder-editor, go into `packages/dashbuilder-editor` and run pnpm start. Then on `localhost:9001` you can edit use the YAML content above and test the component.

### CODE REVIEW

File: packages/dashbuilder-component-map/dev-webapp/webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
```
Comment (paulovmr) [inline packages/dashbuilder-component-map/dev-webapp/webpack.config.js]:
```suggestion
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
```

File: packages/dashbuilder-component-map/jest.config.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
```
Comment (paulovmr) [inline packages/dashbuilder-component-map/jest.config.js]:
```suggestion
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
```

File: packages/dashbuilder-component-map/static/index.html [hunk 3]
```diff
@@ -0,0 +1,47 @@
+<!--
+  ~ Copyright 2020 Red Hat, Inc. and/or its affiliates.
```
Comment (paulovmr) [inline packages/dashbuilder-component-map/static/index.html]:
```suggestion
  ~ Copyright 2022 Red Hat, Inc. and/or its affiliates.
```

File: packages/dashbuilder-component-map/tests/jest.setup.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
```
Comment (paulovmr) [inline packages/dashbuilder-component-map/tests/jest.setup.ts]:
```suggestion
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
```

**Review by paulovmr — CHANGES_REQUESTED** [review]
Nice, @jesuino ! Only a few files missing copyright or with the wrong year. Also, you are adding a file called `bubble_manifest.dev copy.json`, might have been a typo.

**Comment by jesuino** [comment]
(comment on the wrong issue)

**Comment by jesuino** [comment]
Thanks @paulovmr - modified the files

**Review by paulovmr — APPROVED** [review]

File: packages/dashbuilder-component-map/src/Map.tsx [hunk 5]
```diff
@@ -0,0 +1,162 @@
+import { ComponentController } from "@kie-tools/dashbuilder-component-api";
+import React from "react";
+import { ComposableMap, Geographies, Geography, Marker, ZoomableGroup } from "react-simple-maps";
+import { scaleLinear } from "d3-scale";
+
+const DEFAULT_GEO = "world.geo.json";
+const DEFAULT_BACKGROUND = "lightblue";
+const DEFAULT_STROKE = "darkgray";
+const DEFAULT_FILL = "#F2D0A9 ";
```
Comment (barboras7) [inline packages/dashbuilder-component-map/src/Map.tsx]:
```suggestion
const DEFAULT_FILL = "#F2D0A9";
```

**Comment by barboras7** [comment]
Hi @jesuino,
I wanted to test the feature with some more complex YAML file to try all the possible properties but I probably had some issues with the syntax since I did not succeed. Do you have any sample YAML file to use for more complex testing? :slightly_smiling_face:

**Comment by barboras7** [comment]
Hi William, just wondering if you accidentally overlooked my comment or you don't havy any more complex YAML? I'd like to proceed with testing either by figuring it myself or with some YAML from you. Thanks :slightly_smiling_face: 

> Hi @jesuino, I wanted to test the feature with some more complex YAML file to try all the possible properties but I probably had some issues with the syntax since I did not succeed. Do you have any sample YAML file to use for more complex testing? slightly_smiling_face

**Comment by jesuino** [comment]
Hello @barboras7 

For some reason my comment was lost.  Here's a more complex YAML that test all map features!

```
datasets:
    - uuid: data      
      content: >-
                [
                    ["SP", 1], 
                    ["RJ", 90],
                    ["MG", 5],
                    ["PA", 50],
                    ["AM", 2],
                    ["RS", 100]
                ]
      columns:
          - id: name
            type: label
          - id: value
            type: number
pages:
    - components:
          - settings:
                component: map
                map:
                    geoUrl: https://raw.githubusercontent.com/sjcdigital/geo-json/main/Brasil.json
                    enableZoom: "true"
                    minZoom: "1"
                    maxZoom: "20"
                    zoom: "2"
                    fill: white
                    background: lightgray
                    strokeWidth: "0.2"
                    zoomCenterLat: "-28.749997"
                    zoomCenterLong: "-47.9499962"
                    fillBegin: lightgreen
                    fillEnd: darkgreen
                    geoKey: "UF"
                external:
                    width: 100%
                    height: 800px
                lookup:
                    uuid: data
                    group:
                        - functions:
                              - source: name
                              - source: value
```

**Comment by barboras7** [comment]
Thanks! It works as expected… I did not have the “map” under settings in my yaml, that’s why it did not render properly 😣

**Review by barboras7 — APPROVED** [review]

**Pull request merged**